Lib/typing.py

```diff
@@ -1057,6 +1114,214 @@ def __dir__(self):
+        typevars=(Ts,),     args=(int, str)           => {Ts: (int, str)}
+        typevars=(Ts,),     args=(*tuple[()],)        => {Ts: ()}
+        typevars=(Ts,),     args=(*tuple[int],)       => {Ts: (int,)}
+        typevars=(Ts,),     args=(*tuple[int, str],)  => {Ts: (int, str)}
+        typevars=(Ts,),     args=(*tuple[int, ...],)  => {Ts: (*tuple[int, ...])}
+
+        # ===== A single TypeVar and a single TypeVarTuple =====
+
+        # == Too few args ==
+
+        typevars=(T, Ts),   args=()                  => TypeError
+        typevars=(T, Ts),   args=(*tuple[()],)       => TypeError
+        # Again, this means ">= 0 ints", but we need ">= 1".
+        typevars=(T, Ts),   args=(*tuple[int, ...])  => TypeError
+
+        # == Right number of args ==
+
+        typevars=(T, Ts),   args=(int,)              => {T: int, Ts: ()}
+        typevars=(T, Ts)    args=(int, str)          => {T, int, Ts: (str,)}
+        typevars=(T, Ts),   args=(*tuple[int])       => {T: int, Ts: ()}
+        typevars=(T, Ts),   args=(*tuple[int, str])  => {T: int, Ts: (str,)}
+        typevars=(T, Ts),   args=(int, *tuple[str, ...])
+                                            => {T: int, Ts: (*tuple[str, ...],)}
+        typevars=(T, Ts),   args=(*tuple[int], *tuple[str, ...])
+                                            => {T: int, Ts: (*tuple[str, ...],)}
+    """
+
+    # This would be a pretty complicated algorithm in its full generality.
+    # Fortunately, we can make two observations, which simplify things:
+    # 1. As of PEP 646, there can only be a maximum of one TypeVarTuple.
+    # 2. When considering unpacked arbitrary-length tuples
+    #    like *tuple[int, ...], we can't assign them or any part of them to
+    #    regular TypeVars:
+    #    * We can't assign the whole *tuple[int, ...] to a single TypeVar
+    #      because...well, it represents an arbitrary number of types, and a
+    #      TypeVar only holds exactly one type.
+    #    * We can't assign one of the `int`s to a TypeVar, because we can't
+    #      be sure there'll be any `int`s at all: tuple[int, ...] is a tuple
+    #      of *zero* or more ints.
+
+    if not typevars:
+        return {}
+
+    num_typevartuples = sum(1 for x in typevars if isinstance(x, TypeVarTuple))
+    if num_typevartuples > 1:
+        raise TypeError("At most 1 TypeVarTuple may be used in a type "
+                        f"parameter list, but saw {num_typevartuples}")
+
+    # First, we should go through `args` to replace e.g. *tuple[int] with int.
+    new_args = []
+    for arg in args:
+        if (_is_unpacked_tuple(arg)
+                and not _is_unpacked_arbitrary_length_tuple(arg)):
+            arg_tuple = arg.__args__[0]  # The actual tuple[int]
+            new_args.extend(arg_tuple.__args__)
+        else:
+            new_args.append(arg)
+    args = tuple(new_args)
+
+    # Case 1: typevars does not contain any TypeVarTuples
```

Comment:
nit: This is a really long and complicated function. Can we extract helpers instead of having inline comment such as `Case 1`, `Case 2`, etc.? That would make the logic easier to follow.

Reply:
Good call. Done.